feat(post): egress controller (#277)
BedrockStreaming/tech.bedrockstreaming.com: _posts/2021-10-18-increase-performance-and-stability-by-adding-an-egress-controller.md

---
layout: post
title: "Increase performance and stability by adding an Egress Controller"
description: "How to avoid ErrorPortAllocation on AWS NAT Gateways with PHP applications at scale"
author:
    name: Tanguy Falconnet
avatar:
email:
twitter: 
github: tfalconnet
category:

> tags: [AWS, Kubernetes, PHP, HAProxy]

Reviewer: I think it's a good practice to also add existing tags that are matching.
-> https://tech.bedrockstreaming.com/tags.html

image:

>   feature: /posts/2021-10-18-increase-performance-and-stability-by-adding-an-egress-controller/.jpg

Reviewer: filepath is wrong: `.jpg`

comments: true
language: en
---

# Context

> BedRock is using PHP for almost all of the backend API of our streaming platforms (6Play, RTLMost, Salto, …). We have deployed our applications in AWS on our kops-managed Kubernetes cluster. Each of our applications are behind a CDN for caching purposes (CloudFront, Fastly). Which means every time an application needs to access another API, requests go on the internet to access the latter through CDN.

Reviewer: There's no upper case to the `R` in Bedrock ;)

```suggestion
Bedrock is using PHP for almost all of the backend API of our streaming platforms (6Play, RTLMost, Salto, …). We have deployed our applications in AWS on our kops-managed Kubernetes cluster. Each of our applications are behind a CDN for caching purposes (CloudFront, Fastly). Which means every time an application needs to access another API, requests go on the internet to access the latter through CDN.
```

> BedRock is gradually migrating from onprem infrastructure to cloud infrastructure since 2 years and users activity is continuously growing. Furthemore, we started to see tcp connection errors in our applications.

Reviewer: ```suggestion
Bedrock is gradually migrating from onprem infrastructure to cloud infrastructure since 2 years and users activity is continuously growing. Furthemore, we started to see tcp connection errors in our applications.
```

Reviewer: migration is finished for some time now. we started in 2018 ;)

Also, you are using a single paragraph to say two absolute different things : NAT gateway errors are correlated to intensive TCP connection to the outside of our VPC, not to the number of applications running inside it.
Which means those errors are happening on huge loads (special events) and not because we have many backend APIs

IMHO, both sentences in the same paragraph are confusing

Author: ```suggestion
During special events with huge loads on our platforms, we started to see tcp connection errors from our applications to the outside of our VPC.
```

# ErrorPortAllocation source

After a few investigations, we saw that tcp connection errors were correlated with NAT Gateways ErrorPortAllocation.

![Error Port Allocation Graph](/posts/2021-10-18-increase-performance-and-stability-by-adding-an-egress-controller/error-port-allocation.png)

> Some loadtesting on our plateform, which you may see as "no traffic | huge traffic | no traffic"

Reviewer: is this the legend of the above picture ?

Author: Yes, I'll add it directly in [legend]

> In AWS, NAT Gateways are endpoints allowing us to go outside our AWS network. They have hard limits that can’t be modified :

Reviewer: ```suggestion
In AWS, NAT Gateways are endpoints allowing us to go outside our VPC. They have hard limits that can’t be modified :
```

> A NAT gateway can support up to 55,000 simultaneous connections [...]. If the destination IP address, the destination port, or the protocol (TCP/UDP/ICMP) changes, you can create an additional 55,000 connections. For more than 55,000 connections, there is an increased chance of connection errors due to port allocation errors. [...]
[AWS Documentation](https://docs.aws.amazon.com/vpc/latest/userguide/vpc-nat-gateway.html#nat-gateway-limits)

> Our applications always request the same endpoints : other APIs CDN. Destination port or ip or protocol doesn’t change that much, so we start hitting max connection, resulting in ErrorPortAllocation.

Reviewer: ```suggestion
Our applications always request the same endpoints : other APIs CDN. Destination port, ip or protocol doesn’t change that much, so we start hitting max connections, resulting in ErrorPortAllocation.
```

At the same time, we found a very interesting blogpost : [Impact of using HTTP connection pooling for PHP applications at scale](https://techblog.wikimedia.org/2020/10/26/impact-of-using-http-connection-pooling-for-php-applications-at-scale/), which was a very good coincidence.

> As you can read in this post, PHP applications aren’t able to reuse tcp connections, as PHP processes are continuously dying. This creates latency and CPU waste (TLS negotiation and TCP connection lifecycle) but also overconsumption of tcp connections.

Reviewer: ```suggestion
As you can read in wikipedia's post, PHP applications aren’t able to reuse tcp connections, as PHP processes are not sharing information from a request to another. Recreating new connections on the same endpoints is inefficient: adds latency, wastes CPU (TLS negotiation and TCP connection lifecycle) but also overconsumes tcp connections.
```

![Outgoing Traffic without Egress Controller](/posts/2021-10-18-increase-performance-and-stability-by-adding-an-egress-controller/outgoing-traffic-without-egress-schema.png)

# Outgoing requests optimization

## Egress Controller

We thought that a service mesh may be a little overkill in our case, so we tried to add HAProxy as Kubernetes Egress Controller in our clusters. 

![Outgoing Traffic with Egress Controller](/posts/2021-10-18-increase-performance-and-stability-by-adding-an-egress-controller/outgoing-traffic-with-egress-schema.png)

We configured some apps to send a few outgoing requests to Egress Controller. The latter was configured to do tcp re-use and to forward to desired endpoints.

## Effects

> With this optimization, we don’t encounter ErrorPortAllocation anymore. Requests duration are reduced by 20 to 30%, and apps are consuming less CPU.

Reviewer: you should explain why requests are faster and less CPU consuming ?

Author: ```suggestion
With this optimization, we don’t encounter ErrorPortAllocation anymore. Requests duration are reduced by 20 to 30%, and apps are consuming less CPU. Ressources were spent to instantiate new TLS connection, which is now handled by Egress Controller.
```

![CPU consumption of an app configured to use Egress Controller for DynamoDB requests](/posts/2021-10-18-increase-performance-and-stability-by-adding-an-egress-controller/cpu-metrics.png)

# Detailed configuration

> Keep in mind that HAProxy is unable to define backend servers by itself. You must define each needed domain in a frontend and map it with a backend containing one or more servers.

Reviewer: is it really for the reader of the blogpost to keep that in mind or is it to you to explain this subtlety ?

Reviewer: thus, you're directly describing a detail of the end configuration without explaining globally how to implement an egress controller

Reviewer: I'd simply remove that

We generally prefer to use what already exists rather than starting from scratch, so we tried to see if [HAProxy Kubernetes Ingress Controller](https://www.haproxy.com/documentation/kubernetes/latest/installation/community/) could be used as egress.

HAProxy Ingress Controller loads its frontend domains in [Ingress](https://kubernetes.io/docs/concepts/services-networking/ingress/) ressource, and loads backend servers in the associated [Service](https://kubernetes.io/docs/concepts/services-networking/service/) ressource. To inject an external domain as a backend server, we have to use [Service ExternalName](https://kubernetes.io/docs/concepts/services-networking/service/#externalname).

![Detailed Configuration Schema](/posts/2021-10-18-increase-performance-and-stability-by-adding-an-egress-controller/detailed-configuration.png)

> In our case, it was important that frontend domain match the external domains. HAProxy does not replace host header when forwarding the request, and it causes issues with CDNs if host header does not match with requested API.

Reviewer: not clear enough for non-initiated people

```yaml
---
apiVersion: v1
kind: Service
metadata:
  name: app1
spec:
  type: ExternalName
  externalName: app1.example.com
  ports:
    - name: https
      protocol: TCP
      port: 443
      targetPort: 443
---
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: app1
annotations:
    haproxy.org/server-ssl: "true"
    haproxy.org/backend-config-snippet: |
      # See this article for the deep reasons of both parameters: https://www.haproxy.com/fr/blog/http-keep-alive-pipelining-multiplexing-and-connection-pooling/
      # enforce sni with the Host string instead of the 'Host' header, because HAProxy cannot reuse connections with a non-fixed Host SNI value.

>       default-server check-sni <HOST> sni str(<HOST>) resolvers mydns resolve-prefer ipv4

Reviewer: you must replace <HOST> by the exact value :wink:

Author: ```suggestion
      default-server check-sni app1.example.com sni str(app1.example.com) resolvers mydns resolve-prefer ipv4
```

      # make HAProxy reuse connections, because the default safe mode reuses connections only for the same source.ip
      http-reuse always

spec:
  rules:
  - host: app1.example.com
    http:
      paths:
      - backend:
          serviceName: app1
          servicePort: 443
```

When everything is in ready, you will be able to send requests like :

```yaml
curl -H “host: app1.example.com” https://haproxy-egress.default.svc.cluster.local/health
```

> ⚠️ HAProxy Kubernetes Ingress Controller resolves domain names only when it reloads, but it can be fixed by adding a [config snippet](https://www.haproxy.com/documentation/kubernetes/latest/configuration/configmap/#global-config-snippet) :

Reviewer: where do we need to add this snippet ? in which kubernetes object ?

Author: ```suggestion
⚠️ HAProxy Kubernetes Ingress Controller resolves domain names only when it reloads, but it can be fixed by adding a [config snippet to Egress Controller configuration](https://www.haproxy.com/documentation/kubernetes/latest/configuration/configmap/#global-config-snippet) :
```

```yaml
global-config-snippet: |
  resolvers mydns
    nameserver local <MY_DNS>:53
```

# Conclusions
It seems unusual to gain in requests latency by adding a hop in a network. 

> The main problem with this approach is the fact that we are effectively creating a new SPOF (Single Point of Failure) in our clusters if we choose to send all our egress traffic through it. Instead, we are carefully selecting applications that use it, in order to avoid a giant outage of the platform. Some applications are tightly tied to external services and would massively gain from this and others would only be less resilient.

Reviewer: ```suggestion
The main problem with this approach is the fact that we are effectively creating a Single Point of Failure in our clusters if we choose to send all our egress traffic through it. Instead, we are carefully selecting applications that should use an egress controller, in order to refine the configuration little by little. Some applications are tightly tied to external services and would massively gain from this and others would only be less resilient.
```

---
layout: post
title: "Increase performance and stability by adding an Egress Controller in a cluster Kubernetes at AWS"
description: "How to avoid ErrorPortAllocation on AWS NAT Gateways, with PHP applications, at scale"
author:
    name: Tanguy Falconnet
avatar:
email:
twitter: 
github: tfalconnet
category:
tags: [php, aws, cloud, performance, sysadmin, kubernetes, HAProxy]
comments: true
image:
  feature: posts/bonnes-pratiques-web/bedrock.jpg
language: en
---

# Introduction

We recently encountered issues with our PHP applications at scale in our Kubernetes clusters at AWS. We will explain the root cause of these issues, how we fixed them with Egress Controller, and overall improvements. We also added a detailed configuration to use HAProxy as Egress Controller.

# Context

> Bedrock is using PHP for almost all of the backend API of our streaming platforms (6Play, RTLMost, Salto, …). We have deployed our applications in AWS on our kops-managed Kubernetes cluster. Each of our applications is behind a CDN for caching purposes (CloudFront, Fastly). Which means every time an application needs to access another API, requests go on the internet to access the latter through CDN.

Author: ```suggestion
Bedrock is using PHP for almost all of the backend API of our streaming platforms (6Play, RTLMost, Salto, …). We have deployed our applications in AWS on our kops-managed Kubernetes clusters. Each of our applications is behind a CDN for caching purposes (CloudFront, Fastly). This means every time an application needs to access another API, requests go on the internet to access the latter through CDN.
```

> During special events with huge loads on our platforms, we started to see tcp connection errors from our applications to the outside of our VPC.

Author: ```suggestion
During special events with huge loads on our platforms, we started to see TCP connection errors from our applications to the outside of our VPC.
```

# ErrorPortAllocation source

After a few investigations, we saw that TCP connection errors were correlated with NAT Gateways ErrorPortAllocation.

> ![Some loadtesting on our plateform, which you may see as: no traffic, huge traffic, then no traffic again](/images/posts/2021-10-18-increase-performance-and-stability-by-adding-an-egress-controller/error-port-allocation.png)

Author: ```suggestion
![Some loadtesting on our platform, which you may see as no traffic, huge traffic, then no traffic again](/images/posts/2021-10-18-increase-performance-and-stability-by-adding-an-egress-controller/error-port-allocation.png)
```

In AWS, NAT Gateways are endpoints allowing us to go outside our VPC. They have hard limits that can’t be modified:
> A NAT gateway can support up to 55,000 simultaneous connections [...]. If the destination IP address, the destination port, or the protocol (TCP/UDP/ICMP) changes, you can create an additional 55,000 connections. For more than 55,000 connections, there is an increased chance of connection errors due to port allocation errors. [AWS Documentation](https://docs.aws.amazon.com/vpc/latest/userguide/vpc-nat-gateway.html#nat-gateway-limits)

> Our applications always request the same endpoints: other APIs CDN. Destination port, ip or protocol doesn’t change that much, so we start hitting max connections, resulting in ErrorPortAllocation.

Author: ```suggestion
Our applications always request the same endpoints: other APIs CDN. Destination port, IP or protocol doesn’t change that much, so we start hitting max connections, resulting in ErrorPortAllocation.
```

> At the same time, we found a very interesting blogpost: [Impact of using HTTP connection pooling for PHP applications at scale](https://techblog.wikimedia.org/2020/10/26/impact-of-using-http-connection-pooling-for-php-applications-at-scale/), which was a very good coincidence.

Author: ```suggestion
At the same time, we found a very interesting blog post: [Impact of using HTTP connection pooling for PHP applications at scale](https://techblog.wikimedia.org/2020/10/26/impact-of-using-http-connection-pooling-for-php-applications-at-scale/), which was a very good coincidence.
```

As you can read in Wikimedia's post, PHP applications aren’t able to reuse TCP connections, as PHP processes are not sharing information from a request to another. Recreating new connections on the same endpoints is inefficient: adds latency, wastes CPU (TLS negotiation and TCP connection lifecycle) but also overconsumes TCP connections.

> 

Author: Yeah, image title was legend in google doc. But I thought adding a new legend will do to much, so title would be enough. But I can rework that.

![Outgoing Traffic without Egress Controller](/images/posts/2021-10-18-increase-performance-and-stability-by-adding-an-egress-controller/outgoing-traffic-without-egress-schema.png)

# Outgoing requests optimization

## Egress Controller 

> HAproxy is fast and reliable. We use it often and know it well. We already have it as Ingress Controller in our clusters and we know service mesh needs time to be production ready. So we thought a service mesh might be overkill in our case and we tried to add HAProxy as Kubernetes Egress Controller in our clusters. 

Author: ```suggestion
HAproxy is fast and reliable. We use it often and know it well. We already have it as Ingress Controller in our clusters and we know service mesh needs time to be production-ready. So we thought a service mesh might be overkill in our case and we tried to add HAProxy as Kubernetes Egress Controller in our clusters. 
```

![Outgoing Traffic with Egress Controller](/images/posts/2021-10-18-increase-performance-and-stability-by-adding-an-egress-controller/outgoing-traffic-with-egress-schema.png)

We configured some applications to send a few outgoing requests to Egress Controller. The latter was configured to do TCP re-use and to forward to desired endpoints. 

## Effects

> With this optimization, we don’t encounter ErrorPortAllocation anymore. Requests duration are reduced by 20 to 30%, and apps are consuming less CPU. Ressources were spent to instantiate new TLS connection, which is now handled by Egress Controller.

Author: ```suggestion
With this optimization, we don’t encounter ErrorPortAllocation anymore. Requests duration are reduced by 20 to 30%, and apps are consuming less CPU. Ressources were spent to instantiate a new TLS connection, which is now handled by Egress Controller.
```

![CPU consumption of an app configured to use Egress Controller for DynamoDB requests](/images/posts/2021-10-18-increase-performance-and-stability-by-adding-an-egress-controller/cpu-metrics.png)

# Detailed configuration

We generally prefer to use what already exists rather than starting from scratch, so we tried to see if [HAProxy Kubernetes Ingress Controller](https://www.haproxy.com/documentation/kubernetes/latest/installation/community/) could be used as egress.

> HAProxy Ingress Controller loads its frontend domains in [Ingress](https://kubernetes.io/docs/concepts/services-networking/ingress/) ressource, and loads backend servers in the associated [Service](https://kubernetes.io/docs/concepts/services-networking/service/) ressource. To inject an external domain as a backend server, we have to use [Service ExternalName](https://kubernetes.io/docs/concepts/services-networking/service/#externalname).

Author: ```suggestion
HAProxy Ingress Controller loads its frontend domains in [Ingress](https://kubernetes.io/docs/concepts/services-networking/ingress/) resource, and loads backend servers in the associated [Service](https://kubernetes.io/docs/concepts/services-networking/service/) resource. To inject an external domain as a backend server, we have to use [Service ExternalName](https://kubernetes.io/docs/concepts/services-networking/service/#externalname).
```

![Detailed Configuration Schema](/images/posts/2021-10-18-increase-performance-and-stability-by-adding-an-egress-controller/detailed-configuration.png)

> To use HAProxy Kubernetes Ingress Controller as an Egress Controller, we will use Ingress Kubernetes ressource as Egress to define domains handled by the Controller.

Author: ```suggestion
To use HAProxy Kubernetes Ingress Controller as an Egress Controller, we will use Ingress Kubernetes resource as Egress to define domains handled by the Controller.
```

```yaml
---
apiVersion: v1
kind: Service
metadata:
  name: app1
spec:
  type: ExternalName
  externalName: app1.example.com
  ports:
    - name: https
      protocol: TCP
      port: 443
      targetPort: 443
---
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: app1
annotations:
    haproxy.org/server-ssl: "true"
    haproxy.org/backend-config-snippet: |
      # See this article for the deep reasons of both parameters: https://www.haproxy.com/fr/blog/http-keep-alive-pipelining-multiplexing-and-connection-pooling/

>       # enforce sni with the Host string instead of the 'Host' header, because HAProxy cannot reuse connections with a non-fixed Host SNI value.

Author: ```suggestion
      # enforce SNI with the Host string instead of the 'Host' header, because HAProxy cannot reuse connections with a non-fixed Host SNI value.
```

      default-server check-sni app1.example.com sni str(app1.example.com) resolvers mydns resolve-prefer ipv4
      # make HAProxy reuse connections, because the default safe mode reuses connections only for the same source.ip
      http-reuse always

spec:
  rules:
  - host: app1.example.com
    http:
      paths:
      - backend:
          serviceName: app1
          servicePort: 443
```

When everything is ready, you will be able to send requests:

```yaml
curl -H “host: app1.example.com” https://haproxy-egress.default.svc.cluster.local/health
```

> ⚠️ HAProxy Kubernetes Ingress Controller resolves domain names only when it reloads. It means that if a DNS record is updated (IP changed), you may not see it until HAProxy reloads. Reloading happens when a new server is added to a backend for example, resulting in several seconds/minutes without the correct IP to access an external service. But it can be fixed by adding a [config snippet to Egress Controller configuration](https://www.haproxy.com/documentation/kubernetes/latest/configuration/configmap/#global-config-snippet):

Author: Pascal asks me the opposite :)

```yaml
global-config-snippet: |
  resolvers mydns
    nameserver local <MY_DNS>:53
```

# Conclusions

It seems surprising to reduce requests latency by adding a hop in a network. But it does really work, even if it has some limits.

> The main problem with this approach is the fact that we are effectively creating a Single Point of Failure in our clusters if we choose to send all our egress traffic through it. Instead, we are carefully selecting applications that should use an egress controller, in order to refine the configuration little by little. Some applications are tightly tied to external services and would massively gain from this and others would only be less resilient.

Author: ```suggestion
The main problem with this approach is the fact that we are effectively creating a Single Point of Failure in our clusters if we choose to send all our egress traffic through it. Instead, we are carefully selecting applications that should use an egress controller to refine the configuration little by little. Some applications are tightly tied to external services and would massively gain from this and others would only be less resilient.
```
